app/scripts/controllers/metametrics-controller.test.ts

```diff
@@ -0,0 +1,1663 @@
+  app: {
+    name: 'MetaMask Extension',
+    version: VERSION,
+    extensionId: MOCK_EXTENSION_ID,
+  },
+  page: METAMETRICS_BACKGROUND_PAGE_OBJECT,
+  referrer: undefined,
+  userAgent: window.navigator.userAgent,
+  marketingCampaignCookieId: null,
+};
+
+const DEFAULT_SHARED_PROPERTIES = {
+  chain_id: NETWORK_CLIENT_ID_1_CHAIN_ID,
+  locale: LOCALE.replace('_', '-'),
+  environment_type: 'background',
+};
+
+const DEFAULT_EVENT_PROPERTIES = {
+  category: 'Unit Test',
+  extensionId: MOCK_EXTENSION_ID,
+  ...DEFAULT_SHARED_PROPERTIES,
+};
+
+const DEFAULT_PAGE_PROPERTIES = {
+  ...DEFAULT_SHARED_PROPERTIES,
+};
+
+const SAMPLE_PERSISTED_EVENT = {
+  id: 'testid',
+  persist: true,
+  category: 'Unit Test',
+  successEvent: 'sample persisted event success',
+  failureEvent: 'sample persisted event failure',
+  properties: {
+    test: true,
+  },
+};
+
+const SAMPLE_NON_PERSISTED_EVENT = {
+  id: 'testid2',
+  persist: false,
+  category: 'Unit Test',
+  successEvent: 'sample non-persisted event success',
+  failureEvent: 'sample non-persisted event failure',
+  uniqueIdentifier: 'sample-non-persisted-event',
+  properties: {
+    test: true,
+  },
+};
+
+describe('MetaMetricsController', function () {
+  describe('constructor', function () {
+    it('should properly initialize', async function () {
+      const spy = jest.spyOn(segmentMock, 'track');
+      await withController(({ controller }) => {
+        expect(controller.version).toStrictEqual(VERSION);
+        expect(controller.chainId).toStrictEqual(NETWORK_CLIENT_ID_1_CHAIN_ID);
+        expect(controller.state.participateInMetaMetrics).toStrictEqual(true);
+        expect(controller.state.metaMetricsId).toStrictEqual(
+          TEST_META_METRICS_ID,
+        );
+        expect(controller.state.marketingCampaignCookieId).toStrictEqual(null);
+        expect(controller.locale).toStrictEqual(LOCALE.replace('_', '-'));
+        expect(controller.state.fragments).toStrictEqual({
+          testid: SAMPLE_PERSISTED_EVENT,
+        });
+        expect(spy).toHaveBeenCalledTimes(1);
+        expect(spy).toHaveBeenCalledWith(
+          {
+            event: 'sample non-persisted event failure',
+            userId: TEST_META_METRICS_ID,
+            context: DEFAULT_TEST_CONTEXT,
+            properties: {
+              ...DEFAULT_EVENT_PROPERTIES,
+              test: true,
+            },
+            messageId: 'sample-non-persisted-event-failure',
+            timestamp: new Date(),
+          },
+          spy.mock.calls[0][1],
+        );
+      });
+    });
+
+    it('should update when network changes', async function () {
+      await withController(({ controller, triggerNetworkDidChange }) => {
+        triggerNetworkDidChange({
+          networkConfigurationsByChainId: {},
+          selectedNetworkClientId: 'selectedNetworkClientId2',
+          networksMetadata: {},
+        });
+
+        expect(controller.chainId).toStrictEqual(NETWORK_CLIENT_ID_2_CHAIN_ID);
```

Comment:
Looks fine I will adopt the pattern

Reply:
I initially considered importing the files but decided against it. Instead, I applied your technique with custom logic to add only the chainId. Let me know if this approach works for you. I didn’t import the function since I wasn’t certain it was necessary, as I had to import FakeProvider and FakeBlockTracker. too It felt like too many imports for something we can customize simply by returning a chainId, which is the only required element from the config.

Reply:
Gotcha, makes sense. That works for me. Thanks!